Mathlib/Dynamics/TopologicalEntropy/CoverEntropy.lean

```diff
@@ -0,0 +1,617 @@
+    {s : Set X} (h' : IsDynCoverOf T F U n s) :
+    s.Nonempty := by
+  rcases nonempty_biUnion.1 (Nonempty.mono h' h) with ⟨x, x_s, _⟩
+  exact nonempty_of_mem x_s
+
+lemma isDynCoverOf_zero (T : X → X) (F : Set X) (U : Set (X × X)) {s : Set X} (h : s.Nonempty) :
+    IsDynCoverOf T F U 0 s := by
+  simp only [IsDynCoverOf, ball, dynEntourage, not_lt_zero', Prod.map_iterate,
+    iInter_of_empty, iInter_univ, preimage_univ]
+  rcases h with ⟨x, x_s⟩
+  exact subset_iUnion₂_of_subset x x_s (subset_univ F)
+
+lemma isDynCoverOf_univ (T : X → X) (F : Set X) (n : ℕ) {s : Set X} (h : s.Nonempty) :
+    IsDynCoverOf T F univ n s := by
+  simp only [IsDynCoverOf, ball, dynEntourage, Prod.map_iterate, preimage_univ,
+    iInter_univ, iUnion_coe_set]
+  rcases h with ⟨x, x_s⟩
+  exact subset_iUnion₂_of_subset x x_s (subset_univ F)
+
+lemma IsDynCoverOf.nonempty_inter {T : X → X} {F : Set X} {U : Set (X × X)} {n : ℕ} {s : Finset X}
+    (h : IsDynCoverOf T F U n s) :
+    ∃ t : Finset X, IsDynCoverOf T F U n t ∧ t.card ≤ s.card
+    ∧ ∀ x ∈ t, ((ball x (dynEntourage T U n)) ∩ F).Nonempty := by
+  classical
+  use Finset.filter (fun x : X ↦ ((ball x (dynEntourage T U n)) ∩ F).Nonempty) s
+  simp only [Finset.coe_filter, Finset.mem_filter, and_imp, imp_self, implies_true, and_true]
+  apply And.intro _ (Finset.card_mono (Finset.filter_subset _ s))
+  intros y y_F
+  specialize h y_F
+  simp only [Finset.coe_sort_coe, mem_iUnion, Subtype.exists, exists_prop] at h
+  rcases h with ⟨z, z_s, y_Bz⟩
+  simp only [coe_setOf, mem_setOf_eq, mem_iUnion, Subtype.exists, exists_prop]
+  use z
+  exact ⟨⟨z_s, nonempty_of_mem ⟨y_Bz, y_F⟩⟩, y_Bz⟩
+
+/-This lemma is the first step in a submultiplicative-like property of `coverMincard`, with
+  far-reaching consequences such as explicit bounds for the topological entropy
+  (`coverEntropyInfUni_le_card_div`) and an equality between two notions of topological entropy
+  (`coverEntropyInf_eq_coverEntropySup_of_inv`).-/
+lemma IsDynCoverOf.iterate_le_pow {T : X → X} {F : Set X} (F_inv : MapsTo T F F) {U : Set (X × X)}
+    (U_symm : SymmetricRel U) {m : ℕ} (n : ℕ) {s : Finset X} (h : IsDynCoverOf T F U m s) :
+    ∃ t : Finset X, IsDynCoverOf T F (U ○ U) (m * n) t ∧ t.card ≤ s.card ^ n := by
+  classical
+  rcases F.eq_empty_or_nonempty with (rfl | F_nemp)
+  · use ∅; simp
+  have _ : Nonempty X := nonempty_of_exists F_nemp
+  have s_nemp := h.nonempty F_nemp
+  rcases F_nemp with ⟨x, x_F⟩
+  rcases m.eq_zero_or_pos with (rfl | m_pos)
+  · use {x}
+    simp only [zero_mul, Finset.coe_singleton, Finset.card_singleton]
+    exact And.intro (isDynCoverOf_zero T F (U ○ U) (singleton_nonempty x))
+      <| one_le_pow_of_one_le' (Nat.one_le_of_lt (Finset.Nonempty.card_pos s_nemp)) n
+  have : ∀ β : Fin n → s, ∃ y : X, (⋂ k : Fin n, T^[m * k] ⁻¹' ball (β k) (dynEntourage T U m)) ⊆
+      ball y (dynEntourage T (U ○ U) (m * n)) := by
+    intro t
```

Comment:
```suggestion
  have (t : Fin n → s) : ∃ y : X, (⋂ k : Fin n, T^[m * k] ⁻¹' ball (t k) (dynEntourage T U m)) ⊆
      ball y (dynEntourage T (U ○ U) (m * n)) := by
```
`have` lets you write parameters on the left of the colon, to improve readability (and you don't need to `intro` it in the proof).

Reply:
Done. Thanks.